torch/csrc/distributed/c10d/comm.h

```diff
@@ -68,16 +68,39 @@ class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
 
   ~PythonCommHook() override;
 
-  c10::intrusive_ptr<torch::jit::Future> runHook(
-      const GradBucket& bucket) override;
+  c10::intrusive_ptr<torch::jit::Future> runHook(GradBucket& bucket) override;
 
   std::vector<at::Tensor> parseHookResult(const c10::IValue& result) override;
 
  private:
   // Only needed for stateful communication.
   py::object state_;
-  // Indicates an asynchrounous communication of gradients.
   py::object hook_;
 };
 
+class TORCH_API CppCommHook : public CommHookInterface {
+ public:
+  explicit CppCommHook(
+      std::function<c10::intrusive_ptr<
+          torch::jit::Future>(ProcessGroup*, GradBucket&)>& hook,
```

Comment:
Another option here might be to make this class a template class and do this:

```
template <typename T>
class TORCH_API CppCommHook : public CommHookInterface {
  public:
    explicit CppCommHook(
      std::function<c10::intrusive_ptr<
          torch::jit::Future>(T, GradBucket&)>& hook,
```

This way users can use whatever object they want to as their state, for example:

```class FP16Compress : public CommHookInterface<std::shared_ptr<ProcessGroup>> {}```

Reply:
Considered template class before. One minor concern is that, if the state is null, there isn't a nice way to represent this with T, as T can be anything.

I chose ProcessGroup* mainly because I have only seen the cases that the state here is either a process group or null. On the other hand, I cannot use std::optional<ProcessGroup>, optional cannot be applied to an abstract class.

Reply:
We can pass `nullptr` for T in case the state is null? And the subclass could be defined as:

`class MyCompress : public CommHookInterface<void*> {}`

Reply:
Fair enough. Let's use void* to represent a null state.